### DISCUSSION TITLE
Custom stdin/stdout ?

### DISCUSSION THREAD

**Discussion by KaKi87** [esatterwhite/node-seeli#123] [Q&A] [answered]
Hello,

Is it possible to use seeli with other input/output streams than `process.stdin` & `process.stdout` ?

Thanks

**Comment by esatterwhite** [comment, marked as answer]
It depends on what you want to do I suppose. Admittedly, it was not something that I built for in the early days.

Right now, if you want to sent output to something other than `stdout` you would have to handle this in a command. If your command returns something other than a string, seeli will not try to write it to stdout.  The the entirety of that logic is here

https://github.com/esatterwhite/node-seeli/blob/a9d7caf0a2df99372ce2e110b6b3a213600644cb/lib/seeli.js#L89-L100

Instead it will be handed to a content handler in which you can do something custom with your data

```javascript
new Command({
  onContent: function(content) {
    process.stderr.write(content.foo)
  }
, run: async function(flags) {
    return {foo: 'hello world'}
  }
})
```


Its not impossible to change some of the internals of `Command` to take an arbitrary stream object.  Currently, It isn't possible to pipe directly to a seeli command that I can think of. 

```
cat 'hello world' | mycmd caps
$ HELLO WORLD
```
It should be - I agree. 

If you have ideas, want to hash it out, or want to make an attempt at it, more than happy to help.

**Reply by esatterwhite** [reply]
Another problem that I would like to fix, is the internal `ui` object is an ora instance which has its own streams. This is a problem in testing and is kind of clunky to work around. I would also like to fix that and/or make it easier to turn that off

https://github.com/esatterwhite/node-seeli/blob/a9d7caf0a2df99372ce2e110b6b3a213600644cb/lib/command/index.js#L170-L175

**Reply by KaKi87** [reply]
I see, in the end I can just print wherever I want manually inside `onContent` or `run` as long as I don't use `this.ui`.

I guess I didn't take enough time to read the example code on the README, as I've been struggling with many CLI frameworks on that issue already and lost my patience, sorry.

Thanks

**Comment by esatterwhite** [comment]
the internal `Command` is a class which you can sub class and bend to your will for the most part. if examples would help, I can dig some up

as long as it has `run` function that returns a promise - anything is fair game really.
